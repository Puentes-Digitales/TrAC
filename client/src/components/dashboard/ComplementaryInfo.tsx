import { AnimatePresence, motion } from "framer-motion";
import React, { FC, memo, useContext, useEffect, useState } from "react";

import { Flex, Stack, Text } from "@chakra-ui/react";

import { ConfigContext } from "../../context/Config";
import { setTrackingData, track } from "../../context/Tracking";
import { IndicatorTooltip } from "../IndicatorTooltip";

export const ComplementaryInfo: FC<{
  employed?: boolean | null;
  educational_system?: boolean | null;
  institution?: string | null;
  months_to_first_job?: number | null;
  type_admission?: string | null;
  graduation_term?: string | null;
}> = memo(
  ({
    employed,
    educational_system,
    institution,
    months_to_first_job,
    type_admission,
    graduation_term,
  }) => {
    const {
      NO_INFORMATION,
      EMPLOYED_TRUE,
      EMPLOYED_FALSE,
      EDUCATIONAL_SYSTEM_TRUE,
      EDUCATIONAL_SYSTEM_FALSE,
      COMPLEMENTARY_INFORMATION,
      COMPLEMENTARY_INFORMATION_EMPLOYED_TOOLTIP,
      COMPLEMENTARY_INFORMATION_EMPLOYED,
      COMPLEMENTARY_INFORMATION_BACKGROUND_COLOR,
      COMPLEMENTARY_INFORMATION_EMPLOYED_EDUCATIONAL_SYSTEM,
      COMPLEMENTARY_INFORMATION_EMPLOYED_INSTITUTION,
      COMPLEMENTARY_INFORMATION_EMPLOYED_MONTHS,
      COMPLEMENTARY_INFORMATION_GRADUATION_TERM,
      COMPLEMENTARY_INFORMATION_GRADUATION_TERM_TOOLTIP,
      COMPLEMENTARY_INFORMATION_UNGRADUATE,
      COMPLEMENTARY_INFORMATION_TEXT_COLOR,
      COMPLEMENTARY_INFORMATION_TYPE_ADMISSION,
      COMPLEMENTARY_INFORMATION_TYPE_ADMISSION_TOOLTIP,
      COMPLEMENTARY_INFORMATION_EMPLOYED_EDUCATIONAL_SYSTEM_TOOLTIP,
      COMPLEMENTARY_INFORMATION_EMPLOYED_INSTITUTION_TOOLTIP,
      COMPLEMENTARY_INFORMATION_EMPLOYED_MONTHS_TOOLTIP,
    } = useContext(ConfigContext);

    const [show, setShow] = useState(false);

    useEffect(() => {
      setTrackingData({
        showingStudentComplementaryInformation: show,
      });
    }, [show]);

    return (
      <Flex alignItems="center" ml="1em">
        <Flex
          id="complementary_information"
          backgroundColor={COMPLEMENTARY_INFORMATION_BACKGROUND_COLOR}
          boxShadow={
            show
              ? "0px 0px 2px 1px rgb(174,174,174)"
              : "2px 3px 2px 1px rgb(174,174,174)"
          }
          borderRadius={show ? "5px 5px 5px 5px" : "0px 5px 5px 0px"}
          alignItems="center"
          onClick={() => {
            setShow((show) => !show);
            track({
              action: "click",
              effect: show
                ? "close-studentComplementaryInfo"
                : "open-studentComplementaryInfo",
              target: "studentComplementaryInfo",
            });
          }}
          color={COMPLEMENTARY_INFORMATION_TEXT_COLOR}
          cursor="pointer"
          transition="box-shadow 0.2s ease-in-out"
          data-testid="BoxContainer"
        >
          <Stack
            alignItems="center"
            className="unselectable"
            isInline
            pt={10}
            pb={10}
            fontFamily="Lato"
          >
            <Text
              minWidth="90px"
              height="150px"
              m={0}
              ml={4}
              textAlign="center"
              fontWeight="bold"
              className="verticalText"
              fontSize="1.1em"
            >
              {COMPLEMENTARY_INFORMATION}
            </Text>
            <AnimatePresence>
              {show && (
                <motion.div
                  key="employed-text"
                  initial={{
                    opacity: 0,
                  }}
                  animate={{ opacity: 1 }}
                  exit={{
                    opacity: 0,
                  }}
                >
                  {type_admission && (
                    <Flex minWidth="max-content" pr={5} pl={0} pt={1} pb={1}>
                      <Text>
                        {COMPLEMENTARY_INFORMATION_TYPE_ADMISSION}{" "}
                        {type_admission}
                      </Text>
                      <IndicatorTooltip
                        tooltipShow={
                          COMPLEMENTARY_INFORMATION_TYPE_ADMISSION_TOOLTIP
                        }
                      ></IndicatorTooltip>
                    </Flex>
                  )}
                  <Flex minWidth="max-content" pr={5} pl={0} pt={1} pb={1}>
                    <Text>
                      {COMPLEMENTARY_INFORMATION_GRADUATION_TERM}{" "}
                      {graduation_term?.length! > 0
                        ? graduation_term
                        : COMPLEMENTARY_INFORMATION_UNGRADUATE}
                    </Text>
                    <IndicatorTooltip
                      tooltipShow={
                        COMPLEMENTARY_INFORMATION_GRADUATION_TERM_TOOLTIP
                      }
                    ></IndicatorTooltip>
                  </Flex>

                  {typeof employed == "boolean" && (
                    <Flex minWidth="max-content" pr={5} pl={0} pt={1} pb={1}>
                      <Text>
                        {COMPLEMENTARY_INFORMATION_EMPLOYED}{" "}
                        {employed == true ? EMPLOYED_TRUE : EMPLOYED_FALSE}
                      </Text>
                      <IndicatorTooltip
                        tooltipShow={COMPLEMENTARY_INFORMATION_EMPLOYED_TOOLTIP}
                      ></IndicatorTooltip>
                    </Flex>
                  )}
                  {typeof employed == "boolean" && (
                    <Flex minWidth="max-content" pr={5} pl={0} pt={1} pb={1}>
                      <Text>
                        {COMPLEMENTARY_INFORMATION_EMPLOYED_EDUCATIONAL_SYSTEM}{" "}
                        {typeof educational_system == "boolean"
                          ? educational_system == true
                            ? EDUCATIONAL_SYSTEM_TRUE
                            : EDUCATIONAL_SYSTEM_FALSE
                          : NO_INFORMATION}
                      </Text>
                      <IndicatorTooltip
                        tooltipShow={
                          COMPLEMENTARY_INFORMATION_EMPLOYED_EDUCATIONAL_SYSTEM_TOOLTIP
                        }
                      ></IndicatorTooltip>
                    </Flex>
                  )}
                  {typeof employed == "boolean" && (
                    <Flex
                      minWidth="max-content"
                      data-testid="inst"
                      pr={5}
                      pl={0}
                      pt={1}
                      pb={1}
                    >
                      <Text>
                        {COMPLEMENTARY_INFORMATION_EMPLOYED_INSTITUTION}{" "}
                        {typeof institution == "string"
                          ? institution
                          : NO_INFORMATION}
                      </Text>
                      <IndicatorTooltip
                        tooltipShow={
                          COMPLEMENTARY_INFORMATION_EMPLOYED_INSTITUTION_TOOLTIP
                        }
                      ></IndicatorTooltip>
                    </Flex>
                  )}
                  {typeof employed == "boolean" ? (
                    <Flex minWidth="max-content" pr={5} pl={0} pt={1} pb={1}>
                      <Text>
                        {COMPLEMENTARY_INFORMATION_EMPLOYED_MONTHS}{" "}
                        {typeof months_to_first_job == "number"
                          ? months_to_first_job
                          : NO_INFORMATION}
                      </Text>
                      <IndicatorTooltip
                        tooltipShow={
                          COMPLEMENTARY_INFORMATION_EMPLOYED_MONTHS_TOOLTIP
                        }
                      ></IndicatorTooltip>
                    </Flex>
                  ) : null}
                </motion.div>
              )}
            </AnimatePresence>
          </Stack>
        </Flex>
      </Flex>
    );
  }
);

export default ComplementaryInfo;
