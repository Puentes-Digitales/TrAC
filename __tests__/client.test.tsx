import "@testing-library/jest-dom/extend-expect";

import React from "react";
import waitForExpect from "wait-for-expect";

import { MockedProvider } from "@apollo/react-testing";
import {
  act,
  cleanup,
  render,
  screen,
  fireEvent,
} from "@testing-library/react";

import { UserType } from "../client/constants";
import { baseConfig } from "../client/constants/baseConfig";
import { baseUserConfig } from "../client/constants/userConfig";
import {
  AllUsersAdminDocument,
  CheckUnlockDocument,
  CurrentUserDocument,
} from "../client/src/graphql";
import AdminPage from "../client/src/pages/admin";
import LoginPage from "../client/src/pages/login";
import UnlockPage from "../client/src/pages/unlock/[email]/[unlockKey]";
import ComplementaryInfo from "../client/src/components/dashboard/ComplementaryInfo";
import { IndicatorTooltip } from "../client/src/components/IndicatorTooltip";
import { Popover } from "@chakra-ui/react";
Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

afterEach(async () => {
  await cleanup();
});

describe("unlock", () => {
  test("renders correctly", async () => {
    await act(async () => {
      const { getByText } = render(
        <MockedProvider
          mocks={[
            {
              request: {
                query: CurrentUserDocument,
              },
              result: {
                data: {
                  currentUser: {
                    user: null,
                  },
                },
              },
            },
            {
              request: {
                query: CheckUnlockDocument,
                variables: {
                  email: "asd@gmail.com",
                  unlockKey: "XwPp9xazJ0ku5CZnlmgAx2Dld8SHkAeT",
                },
              },
              result: {
                data: {
                  checkUnlockKey: null,
                },
              },
            },
          ]}
          addTypename={false}
        >
          <UnlockPage
            email="asd@gmail.com"
            unlockKey="XwPp9xazJ0ku5CZnlmgAx2Dld8SHkAeT"
          />
        </MockedProvider>
      );

      await waitForExpect(async () => {
        const NewPasswordFieldLabel = getByText(
          baseConfig.UNLOCK_NEW_PASSWORD_LABEL
        );

        expect(NewPasswordFieldLabel).toBeTruthy();
      });
    });
  });
});

describe("login", () => {
  test("renders correctly", async () => {
    await act(async () => {
      const { getByText } = render(
        <MockedProvider
          mocks={[
            {
              request: {
                query: CurrentUserDocument,
              },
              result: {
                data: {
                  currentUser: {
                    user: null,
                  },
                },
              },
            },
          ]}
          addTypename={false}
        >
          <LoginPage />
        </MockedProvider>
      );

      await waitForExpect(async () => {
        const LoginButton = getByText(baseConfig.LOGIN_BUTTON);

        expect(LoginButton).toBeTruthy();
        expect(LoginButton).toHaveAttribute("disabled");
      });
    });
  });
});

describe("admin", () => {
  test("renders correctly", async () => {
    await act(async () => {
      const { getByText } = render(
        <MockedProvider
          mocks={[
            {
              request: {
                query: CurrentUserDocument,
              },
              result: {
                data: {
                  currentUser: {
                    user: {
                      email: "asd@gmail.com",
                      name: "name",
                      admin: true,
                      type: UserType.Director,
                      config: baseUserConfig,
                      student_id: "",
                      __typename: "User",
                    },
                    __typename: "AuthResult",
                  },
                },
              },
            },
            {
              request: {
                query: AllUsersAdminDocument,
              },
              result: {
                data: {
                  users: [],
                },
              },
            },
          ]}
          addTypename={true}
        >
          <AdminPage />
        </MockedProvider>
      );

      await waitForExpect(async () => {
        const UsersMenu = getByText("Users");

        expect(UsersMenu).toBeTruthy();
        expect(UsersMenu).toHaveClass("active");
      });
    });
  });
});

describe("Test <AdmissionDropout />", () => {
  test("Test Snapshot", async () => {
    await act(async () => {
      const { getByTestId } = render(
        <ComplementaryInfo
          type_admission="PSU"
          educational_system={true}
          institution="Escuela"
          months_to_first_job={6}
          graduation_term="2020-2"
        />
      );

      await waitForExpect(() => {
        fireEvent.click(getByTestId("BoxContainer"));
        expect(getByTestId("BoxContainer")).toMatchSnapshot();
      });
    });
  });
});

test("test props en <AdmissionDropout />", async () => {
  const employed = true;
  const type_admission = "PSU";
  const educational_system = true;
  const institution = "Escuela";
  const months_to_first_job = 6;
  const graduation_term = "2022-2";

  await act(async () => {
    const { getByText, getByTestId } = render(
      <ComplementaryInfo
        employed={employed}
        type_admission={type_admission}
        educational_system={educational_system}
        institution={institution}
        months_to_first_job={months_to_first_job}
        graduation_term={graduation_term}
      />
    );
    await waitForExpect(() => {
      fireEvent.click(getByTestId("BoxContainer"));
      expect(getByText("Tipo de ingreso: PSU".trim())).toBeInTheDocument();
      expect(getByText("Sistema educacional: Si".trim())).toBeInTheDocument();
    });
  });
});

test("test null props complementary component", async () => {
  const employed = true;
  const type_admission = null;
  const educational_system = true;
  const institution = null;
  const months_to_first_job = null;
  const graduation_term = null;

  await act(async () => {
    const { getByTestId, queryAllByText } = render(
      <ComplementaryInfo
        employed={employed}
        type_admission={type_admission}
        educational_system={educational_system}
        institution={institution}
        months_to_first_job={months_to_first_job}
        graduation_term={graduation_term}
      ></ComplementaryInfo>
    );
    await waitForExpect(() => {
      fireEvent.click(getByTestId("BoxContainer"));
      const inst_value = queryAllByText("Tipo de ingreso:");
      expect(inst_value).toHaveLength(0);
      const education_system_value = queryAllByText(
        "Sistema educacional: Si".trim()
      );
      expect(education_system_value).toHaveLength(1);
    });
  });
});

describe("indicatorTooltip", () => {
  test("render correctly", async () => {
    await act(async () => {
      const { getByTestId, getByText } = render(
        <IndicatorTooltip tooltipShow="any" />
      );
      await waitForExpect(() => {
        fireEvent.mouseEnter(getByTestId("test-tooltip"));
        expect(getByText("any")).toBeInTheDocument();
      });
    });
  });
});

export {};
