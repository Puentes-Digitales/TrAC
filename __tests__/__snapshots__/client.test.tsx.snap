// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <AdmissionDropout /> Test Snapshot 1`] = `
<div
  class="css-1b6jfu7"
  data-testid="BoxContainer"
  id="complementary_information"
>
  <div
    class="chakra-stack unselectable css-16rokbl"
  >
    <p
      class="chakra-text verticalText css-1568c7u"
    >
      Información Complementaria
    </p>
    <div
      style="opacity: 0;"
    >
      <div
        class="css-1nm7xg4"
      >
        <p
          class="chakra-text css-0"
        >
          Tipo de ingreso: 
           
          PSU
        </p>
        <svg
          aria-controls="popover-content-2"
          aria-expanded="false"
          aria-haspopup="dialog"
          class="chakra-icon css-psva3u"
          data-testid="test-tooltip"
          focusable="false"
          id="popover-trigger-2"
          viewBox="0 0 24 24"
        >
          <g
            stroke="currentColor"
            stroke-width="1.5"
          >
            <path
              d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
              fill="none"
              stroke-linecap="round"
            />
            <path
              d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
              fill="currentColor"
              stroke-linecap="round"
            />
            <circle
              cx="12"
              cy="12"
              fill="none"
              r="11.25"
              stroke-miterlimit="10"
            />
          </g>
        </svg>
      </div>
      <div
        class="css-1nm7xg4"
      >
        <p
          class="chakra-text css-0"
        >
          Fecha de titulación:
           
          2020-2
        </p>
        <svg
          aria-controls="popover-content-4"
          aria-expanded="false"
          aria-haspopup="dialog"
          class="chakra-icon css-psva3u"
          data-testid="test-tooltip"
          focusable="false"
          id="popover-trigger-4"
          viewBox="0 0 24 24"
        >
          <g
            stroke="currentColor"
            stroke-width="1.5"
          >
            <path
              d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
              fill="none"
              stroke-linecap="round"
            />
            <path
              d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
              fill="currentColor"
              stroke-linecap="round"
            />
            <circle
              cx="12"
              cy="12"
              fill="none"
              r="11.25"
              stroke-miterlimit="10"
            />
          </g>
        </svg>
      </div>
    </div>
  </div>
</div>
`;
